### PR TITLE
Lazy loading

### LINKED ISSUE #88: Add lazy loading features
These 3 endpoints support lazy-loading room members:
- GET /sync
- GET /rooms/{roomId}/messages
- GET /rooms/{roomId}/context/{eventId}

I think all that needs to happen so to modify the properties add `lazy_load_members`, `include_redundant_members` to `filter::RoomEventFilter`, but the implementer should look at matrix-org/matrix-doc#2035 for more information.

### PR DESCRIPTION
**Pull request by iinuwa**
Fixes #88.

### CODE REVIEW

**Comment by jplatte** [comment]
To add some more context about my review: As written currently, deserializing a lazy-loading options from `{ "include_redundant_members": true }` wouldn't work because of the missing `lazy_load_members`. I don't see why this shouldn't work per the spec.

What I'd like to see for the boolean fields is

```rust
#[serde(default, skip_serializing_if = "is_false")]
field: bool,
```

so that we accept literal `false` and the field being absent as `false`, but always serialize `false` as an absent field (because that's more efficient and means the same thing).

**Comment by iinuwa** [comment]
> Why create a separate struct `LazyLoadOptions`, if it's only used in one place? Also, if it is a separate struct it will need to be `#[serde(flatten)]` and non-optional, with the individual fields instead having a `#[serde(default)]` (I don't see a reason for them being optional, since `None`, `Some(true)` and `Some(false)` would have three different meanings).

Yeah, that's an oversight on my part due to multi-tasking. When I first started out, I thought I might need to add the `LazyLoadOptions` to each endpoint, but then I realized that I only needed to add it to the `RoomEventFilter`.

I was wondering how to communicate that setting `include_redundant_members` doesn't mean anything unless `lazy_load_members` is true using the type system, but I was having a bit of trouble not making it awkward. I suppose that could just be a comment, and your solution would produce the same message on the wire, so the payload would be the same.

I'll try to make that change this afternoon.

**Comment by jplatte** [comment]
Ah, I actually didn't read the spec carefully enough! Well if `include_redundant_members` only means something with `lazy_load_members = true`, we can represent the possible settings as

```rust
enum LazyLoadingOptions {
    Disabled,
    Enabled { include_redundant_members: bool }
}
```

with custom `Serialize` and `Deserialize` implementations and `#[serde(flatten)]`.

**Comment by iinuwa** [comment]
OK, that makes sense.

**Comment by iinuwa** [comment]
All right, I tried something. Let me know if you have any ideas on improvements.

File: src/r0/filter.rs [hunk 1]
```diff
@@ -203,3 +212,109 @@ impl FilterDefinition {
...
+
+impl Default for LazyLoadOptions {
+    fn default() -> Self {
+        Self::Disabled
+    }
+}
+
+struct LazyLoadOptionsVisitor {}
+
+impl<'de> Visitor<'de> for LazyLoadOptionsVisitor {
```
Comment (jplatte) [inline src/r0/filter.rs]: I have a feeling the visitor stuff here isn't needed, but haven't figured out how yet. To my understanding, the visitor stuff is only really needed when multiple `visit_` methods are implemented  (when multiple types of input are valid).

You could try `deserializer.deserialize_struct`. But I'm also fine merging this without that.

File: src/r0/filter.rs [hunk 2]
```diff
@@ -203,3 +212,109 @@ impl FilterDefinition {
...
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_str("Lazy load options")
+    }
+
+    fn visit_map<M>(self, mut access: M) -> Result<Self::Value, M::Error>
+    where
+        M: MapAccess<'de>,
+    {
+        let mut lazy_load_members = false;
+        let mut include_redundant_members: Option<bool> = Option::None;
```
Comment (jplatte) [inline src/r0/filter.rs]: Why not

```suggestion
        let mut include_redundant_members = false;
```

? AFAICT, that would simplify the logic below a bit.

Comment (iinuwa) [reply src/r0/filter.rs]: Because I wrote this nearing midnight, and my critical thinking skills are shut off after 8 pm. 😂

File: src/r0/filter.rs (hunk 1 above)
Comment (iinuwa) [reply src/r0/filter.rs]: It seems that `deserialize_struct` also [requires a visitor](https://docs.serde.rs/serde/de/trait.Deserializer.html#tymethod.deserialize_struct). Unless `serde` provides some default `Visitor` implementations, I don't think we can get away without it. At least to my understanding.

Comment (jplatte) [reply src/r0/filter.rs]: I see. Sorry for not reading the docs carefully before writing that comment.

File: src/r0/filter.rs [hunk 3]
```diff
@@ -203,3 +212,142 @@ impl FilterDefinition {
...
+                _ => (),
+            };
+        }
+
+        if lazy_load_members {
+            return Ok(LazyLoadOptions::Enabled {
+                include_redundant_members,
+            });
+        }
+        Ok(LazyLoadOptions::Disabled)
```
Comment (jplatte) [inline src/r0/filter.rs]:
```suggestion
        Ok(if lazy_load_members {
            LazyLoadOptions::Enabled {
                include_redundant_members,
            }
        } else {
            LazyLoadOptions::Disabled
        })
```

modulo formatting

File: src/r0/filter.rs [hunk 4]
```diff
@@ -203,3 +212,142 @@ impl FilterDefinition {
...
+
+impl<'de> Deserialize<'de> for LazyLoadOptions {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        deserializer.deserialize_map(LazyLoadOptionsVisitor {})
+    }
+}
+#[cfg(test)]
```
Comment (jplatte) [inline src/r0/filter.rs]:
```suggestion
}

#[cfg(test)]
```

File: src/r0/filter.rs [hunk 5]
```diff
@@ -203,3 +212,142 @@ impl FilterDefinition {
...
+mod tests {
+    use serde_json::json;
+
+    use super::LazyLoadOptions;
+    #[test]
```
Comment (jplatte) [inline src/r0/filter.rs]:
```suggestion
    use super::LazyLoadOptions;

    #[test]
```

File: src/r0/filter.rs [hunk 6]
```diff
@@ -203,3 +212,142 @@ impl FilterDefinition {
...
+    }
+
+    #[test]
+    fn test_serializing_lazy_load_no_redundant() {
+        let lazy_load_options = LazyLoadOptions::Enabled {
+            include_redundant_members: false,
+        };
+        assert_eq!(
+            serde_json::to_value(lazy_load_options).unwrap(),
+            json!({"lazy_load_members": true})
```
Comment (jplatte) [inline src/r0/filter.rs]:
```suggestion
            json!({ "lazy_load_members": true })
```

File: src/r0/filter.rs [hunk 7]
```diff
@@ -203,3 +212,142 @@ impl FilterDefinition {
...
+    }
+
+    #[test]
+    fn test_serializing_lazy_load_with_redundant() {
+        let lazy_load_options = LazyLoadOptions::Enabled {
+            include_redundant_members: true,
+        };
+        assert_eq!(
+            serde_json::to_value(lazy_load_options).unwrap(),
+            json!({"lazy_load_members": true, "include_redundant_members": true})
```
Comment (jplatte) [inline src/r0/filter.rs]:
```suggestion
            json!({ "lazy_load_members": true, "include_redundant_members": true })
```

File: src/r0/filter.rs [hunk 8]
```diff
@@ -203,3 +212,142 @@ impl FilterDefinition {
...
+        );
+    }
+
+    #[test]
+    fn test_deserializing_no_lazy_load() {
+        let input = r#"{"include_redundant_members": true}"#;
```
Comment (jplatte) [inline src/r0/filter.rs]:
```suggestion
        let input = r#"{ "include_redundant_members": true }"#;
```

File: src/r0/filter.rs [hunk 9]
```diff
@@ -203,3 +212,145 @@ impl FilterDefinition {
...
+impl Serialize for LazyLoadOptions {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut state;
+        match *self {
+            Self::Enabled {
+                include_redundant_members,
+            } if include_redundant_members => {
```
Comment (jplatte) [inline src/r0/filter.rs]:
```suggestion
                include_redundant_members: true,
            } => {
```

File: src/r0/filter.rs [hunk 10]
```diff
@@ -203,3 +212,145 @@ impl FilterDefinition {
...
+            }
+            _ => {
+                state = serializer.serialize_struct("LazyLoad", 0)?;
+            }
+        }
+        state.end()
+    }
+}
+
+impl Default for LazyLoadOptions {
```
Comment (jplatte) [inline src/r0/filter.rs]: This is fine, but out of curiosity: Why did you add it?

File: src/r0/filter.rs [hunk 11]
```diff
@@ -203,3 +212,145 @@ impl FilterDefinition {
...
+    fn default() -> Self {
+        Self::Disabled
+    }
+}
+
+struct LazyLoadOptionsVisitor {}
```
Comment (jplatte) [inline src/r0/filter.rs]:
```suggestion
struct LazyLoadOptionsVisitor;
```

File: src/r0/filter.rs [hunk 12]
```diff
@@ -203,3 +212,145 @@ impl FilterDefinition {
...
+
+impl<'de> Visitor<'de> for LazyLoadOptionsVisitor {
+    type Value = LazyLoadOptions;
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
```
Comment (jplatte) [inline src/r0/filter.rs]:
```suggestion
    type Value = LazyLoadOptions;

    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
```

File: src/r0/filter.rs [hunk 13]
```diff
@@ -203,3 +212,145 @@ impl FilterDefinition {
...
+        })
+    }
+}
+
+impl<'de> Deserialize<'de> for LazyLoadOptions {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        deserializer.deserialize_map(LazyLoadOptionsVisitor {})
```
Comment (jplatte) [inline src/r0/filter.rs]:
```suggestion
        deserializer.deserialize_map(LazyLoadOptionsVisitor)
```

File: src/r0/filter.rs (hunk 10 above)
Comment (iinuwa) [reply src/r0/filter.rs]: Oh, that was aleftover from an experiment to see if default implementations would help cut down on some of the serde code. Turned out not to help.

Comment (jplatte) [reply src/r0/filter.rs]: Okay :)
